File: rub/filmgardeCineplexes.rb
# ----------

# FILMGARDE CINEPLEXES SCRAPER

# - parses data from https://fgcineplex.com.sg/movies
# - extracts the following
    # hover-box tour-hotel-box show-product-box 
        # tour-img image
            # a href
            # img src
        # img-category
            # img alt (for the rating)
        # left_side
            # h4
            # a href
        # list-inline
            # Duration, Genre, Language
        # read-btn by-ticket

# ----------

require 'selenium-webdriver'
require 'nokogiri'
require 'json'

def fetch_and_parse_html(url)
  begin
    driver = Selenium::WebDriver.for :chrome
    driver.navigate.to url

    previous_height = driver.execute_script('return document.body.scrollHeight')
    loop do
      driver.execute_script('window.scrollTo(0, document.body.scrollHeight)')
      sleep 3 
      current_height = driver.execute_script('return document.body.scrollHeight')
      break if current_height == previous_height
      previous_height = current_height
    end

    content = driver.page_source
    driver.quit
    content
  rescue => e
    puts "Error fetching HTML: #{e.message}"
    nil
  end
end

def extract_films(url)
  html_text = fetch_and_parse_html(url)
  return unless html_text

  doc = Nokogiri::HTML(html_text)
  fg_films = doc.css('.hover-box.tour-hotel-box.show-product-box')

  def decode_html_entity(encoded_string)
    Nokogiri::HTML::DocumentFragment.parse(encoded_string).text
  end

  def extract_details(details)
    duration_regex = /Duration:\s*(\d+)\s*mins/
    genre_regex = /Genre:\s*([^,]+)/
    language_regex = /Language:\s*(.*)/

    duration_match = details.match(duration_regex)
    genre_match = details.match(genre_regex)
    language_match = details.match(language_regex)

    duration = duration_match ? "#{duration_match[1]} mins" : 'N/A'
    genre = genre_match ? genre_match[1] : 'N/A'
    language = language_match ? language_match[1] : 'N/A'
    [duration, genre, language]
  end

  fg_films_object = []

  if fg_films.any?
    fg_films.each do |fg_film|
      title = fg_film.at_css('.left_side h4 a')&.inner_html || 'N/A'
      title_text = decode_html_entity(title)
      poster_src = fg_film.at_css('.tour-img.image img')&.[]('src') || 'N/A'
      rating = fg_film.at_css('.img-category img')&.[]('alt') || 'N/A'
      book_ticket_url = fg_film.at_css('.read-btn.by-ticket')&.[]('href') || 'N/A'
      raw_details = fg_film.at_css('.list-inline')&.css('li p')
      raw_details_array = raw_details ? raw_details.map { |li| decode_html_entity(li.inner_html).gsub(/<\/?(b|em|strong|i|p)\b[^>]*>/i, '').gsub(/<br\s*\/?>/, ' | ').gsub(/\s{2,}/, ' ').strip }.join(', ') : 'N/A'
      duration, theme, language = extract_details(raw_details_array)

      fg_films_object << {
        title: title_text,
        tags: theme,
        description: language,
        duration: duration,
        rating: rating,
        book_tickets_url: book_ticket_url,
        poster_src: poster_src
      }
    end

    puts "Number of fg movies => #{fg_films_object.length}"
    fg_films_object
  else
    puts 'No filmgarde cineplexes films were found'
    nil
  end
rescue => e
  puts "Error extracting films: #{e.message}"
  nil
end

# ----- EXECUTION CODE -----
  # For testing

films = extract_films("https://fgcineplex.com.sg/movies")
if films
  puts JSON.pretty_generate(films)
else
  puts 'No movies data were found'
end

File: lib/filmgardeCineplexes.js
// ----------

// FILMGARDE CINEPLEXES SCRAPER

// - parses data from https://fgcineplex.com.sg/movies
// - extracts the following
    // hover-box tour-hotel-box show-product-box 
        // tour-img image
            // a href
            // img src
        // img-category
            // img alt (for the rating)
        // left_side
            // h4
            // a href
        // list-inline
            // Duration, Genre, Language
        // read-btn by-ticket

// - dependancy installation
    // npm install playwright
    //  npx playwright install

// ----------

const { chromium } = require('playwright');

async function fetchAndParseHTML(url) {
    try {
        const browser = await chromium.launch();
        const page = await browser.newPage();
        await page.goto(url, { waitUntil: 'networkidle' });
        let previousHeight;
        let currentHeight = await page.evaluate(() => document.body.scrollHeight);
        while (true) {
            previousHeight = currentHeight;
            await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
            await page.waitForTimeout(3000); // wait for content to load
            currentHeight = await page.evaluate(() => document.body.scrollHeight);
            if (currentHeight === previousHeight) break;
        }
        const content = await page.content();
        await browser.close();
        return content;
    } catch (error) {
        console.error('Error fetching HTML:', error);
    }
}

async function extractFilmsFgc(url) {

    const htmlText = await fetchAndParseHTML(url);

    if (htmlText) {
        const { JSDOM } = require('jsdom');
        const dom = new JSDOM(htmlText);
        const doc = dom.window.document;

        // -- debugging -- 
        // console.log(doc.title);
        // console.log(doc.body.innerHTML)

        const fgFilms = doc.querySelectorAll('.hover-box.tour-hotel-box.show-product-box');

        function decodeHTMLEntity(encodedString) {
            const textarea = doc.createElement('textarea');
            textarea.innerHTML = encodedString;
            return textarea.value;
        }

        function extractDetails(details) {
            const durationRegex = /Duration:\s*(\d+)\s*mins/;
            const genreRegex = /Genre:\s*([^,]+)/;
            const languageRegex = /Language:\s*(.*)/;
            const durationMatch = details.match(durationRegex);
            const genreMatch = details.match(genreRegex);
            const languageMatch = details.match(languageRegex);
            const duration = durationMatch ? `${durationMatch[1]} mins` : 'N/A';
            const genre = genreMatch ? genreMatch[1] : 'N/A';
            const language = languageMatch ? languageMatch[1] : 'N/A';
            return [duration, genre, language];
        }

        const fgFilmsObject = [];

        if (fgFilms.length) {
            fgFilms.forEach(fgFilm => {

                    // console.log(fgFilm.innerHTML);

                    const title = fgFilm.querySelector('.left_side h4 a')?.innerHTML || 'N/A';
                    const titleText = decodeHTMLEntity(title);
                    const posterSrc = fgFilm.querySelector('.tour-img.image img')?.getAttribute('src') || 'N/A';
                    const rating = fgFilm.querySelector('.img-category img')?.getAttribute('alt') || 'N/A';
                    const bookTicketUrl = fgFilm.querySelector('.read-btn.by-ticket')?.getAttribute('href') || 'N/A';
                    const rawDetails = fgFilm.querySelector('.list-inline')?.querySelectorAll('li p');
                    const rawDetailsArray = rawDetails ? Array.from(rawDetails).map(li => decodeHTMLEntity(li.innerHTML)?.replace(/<\/?(b|em|strong|i|p)\b[^>]*>/gi, '').replace(/<br\s*\/?>/gi, ' | ').replace(/\s{2,}/g, ' ').trim()).join(', ') : ['N/A'];
                    [duration, theme, language] = extractDetails(rawDetailsArray);

                    // console.log(`
                    //         -----
                    //         Title: ${titleText}
                    //         Theme: ${theme}
                    //         Duration: ${duration}
                    //         Description: ${language}
                    //         Rating: ${rating}
                    //         Poster src: ${posterSrc}
                    //         Book ticket Url: ${bookTicketUrl}
                    //         -----
                    //     `);

                    fgFilmsObject.push({
                        title: titleText,
                        tags: theme,
                        description: language,
                        duration: duration,
                        rating: rating,
                        bookTicketsUrl: bookTicketUrl,
                        posterSrc: posterSrc
                    });

            });

            console.log(`number of fg movies => ${fgFilmsObject.length}`);

            return fgFilmsObject;
        } else {
            console.log('no filmgarde cineplexes films were found');
        }
    } else {
        console.log('HTML content not found');
    }
};

module.exports = { extractFilmsFgc };

// ----- EXECUTION CODE -----
    // - for testing
    // - actual usage to be called from /src/main.js

// extractFilms("https://fgcineplex.com.sg/movies").then((fgFilmsObject) => {
//     if (fgFilmsObject) {
//         console.log(JSON.stringify(fgFilmsObject, null, 2));
//     } else {
//         console.log('no movies data were found');
//     }
// });
